### DISCUSSION TITLE
mTLS support

### DISCUSSION THREAD

**Discussion by afawaz2** [cloudflare/workerd#330] [Q&A] [answered]
We are looking to using workerd to allow trusted third parties to run data mappers in our environment. Some of those mappers might require calling public APIs that use mTLS. Does the fetch API in workerd allow setting client certificates?

**Comment by kentonv** [comment, marked as answer]
As a matter of fact, yes, it does!

You will want to configure a service of type [`ExternalServer`](https://github.com/cloudflare/workerd/blob/2516db7b9dcfa78a82dbe88a11702e8836ec6f12/src/workerd/server/workerd.capnp#L506) which points at the specific remote API endpoint. You can then specify its [`tlsOptions`](https://github.com/cloudflare/workerd/blob/2516db7b9dcfa78a82dbe88a11702e8836ec6f12/src/workerd/server/workerd.capnp#L546), where you would set a [`keypair`](https://github.com/cloudflare/workerd/blob/2516db7b9dcfa78a82dbe88a11702e8836ec6f12/src/workerd/server/workerd.capnp#L710). Since these options are applying on the client side, the keypair will be used as a client certificate when talking to the given service. You would then configure your worker with a service binding pointing at the `ExternalServer`.
